Zend/zend_string.c

```
@@ -72,175 +57,207 @@ ZEND_KNOWN_STRINGS(_ZEND_STR_DSC)
 {
-#ifndef ZTS
 	zend_ulong h;
 	uint32_t nIndex;
 	uint32_t idx;
 	Bucket *p;
 
-	if (ZSTR_IS_INTERNED(str)) {
-		return str;
-	}
-
 	h = zend_string_hash_val(str);
-	nIndex = h | CG(interned_strings).nTableMask;
-	idx = HT_HASH(&CG(interned_strings), nIndex);
+	nIndex = h | interned_strings->nTableMask;
+	idx = HT_HASH(interned_strings, nIndex);
 	while (idx != HT_INVALID_IDX) {
-		p = HT_HASH_TO_BUCKET(&CG(interned_strings), idx);
+		p = HT_HASH_TO_BUCKET(interned_strings, idx);
 		if ((p->h == h) && (ZSTR_LEN(p->key) == ZSTR_LEN(str))) {
 			if (!memcmp(ZSTR_VAL(p->key), ZSTR_VAL(str), ZSTR_LEN(str))) {
-				zend_string_release(str);
 				return p->key;
 			}
 		}
 		idx = Z_NEXT(p->val);
 	}
 
-	GC_REFCOUNT(str) = 1;
-	GC_FLAGS(str) |= IS_STR_INTERNED;
-
-	if (CG(interned_strings).nNumUsed >= CG(interned_strings).nTableSize) {
-		if (CG(interned_strings).nTableSize < HT_MAX_SIZE) {	/* Let's double the table size */
-			void *new_data;
-			void *old_data = HT_GET_DATA_ADDR(&CG(interned_strings));
-			Bucket *old_buckets = CG(interned_strings).arData;
-
-			CG(interned_strings).nTableSize += CG(interned_strings).nTableSize;
-			CG(interned_strings).nTableMask = -CG(interned_strings).nTableSize;
-			new_data = malloc(HT_SIZE(&CG(interned_strings)));
-
-			if (new_data) {
-				HT_SET_DATA_ADDR(&CG(interned_strings), new_data);
-				memcpy(CG(interned_strings).arData, old_buckets, sizeof(Bucket) * CG(interned_strings).nNumUsed);
-				free(old_data);
-				zend_hash_rehash(&CG(interned_strings));
-			} else {
-				CG(interned_strings).nTableSize = CG(interned_strings).nTableSize >> 1;
-				CG(interned_strings).nTableMask = -CG(interned_strings).nTableSize;
-			}
+	return NULL;
+}
+
+static zend_never_inline void zend_string_table_grow(HashTable *interned_strings)
+{
+	if (EXPECTED(interned_strings->nTableSize < HT_MAX_SIZE)) {	/* Let's double the table size */
```

Probably it should fail hard in this situation, as not interning might cause more hard weight issues, at least with TS. Alternatively, we could increase the value, if possible.